Source/DafnyCore/Compilers/Rust/Dafny-compiler-rust-proofs.dfy

```diff
@@ -0,0 +1,362 @@
+    } else {
+      if s[0] == '_' { // "__" + dafnyEscape(s[1..])
+      } else if s[0] == '?' { // "_q" + dafnyEscape(s[1..]) else
+      } else if s[0] == '\'' { // "_k" + dafnyEscape(s[1..]) else
+      } else if s[0] == '#' { // "_h" + dafnyEscape(s[1..]) else=
+      } else { // [s[0]] + dafnyEscape(s[1..])=
+        dafnyEscapeCorrect(s[1..]);
+      }
+    }
+  }
+  // replaceDots is not reversible generally speaking.
+  // Indeed, both _d and . get translated to _d.
+  lemma replaceDotsNotReversible()
+    ensures replaceDots("_d") == replaceDots(".")
+    ensures !isDafnyEncodedIdTail("_d")
+  {}
+
+  // However, if we restrict ourself to a Dafny encoding, it's reversible
+  lemma {:rlimit 500}
+  replaceDotsReversible(i: string)
+    requires isDafnyEncodedIdTail(i)
+    ensures reverseReplaceDots(replaceDots(i)) == i
+  {
+    if |i| == 0 {
+    } else if i[0] == '.' { //"_d" + replaceDots(i[1..])
+      replaceDotsReversible(i[1..]);
+      assert reverseReplaceDots(replaceDots(i)) == i;
+    } else { // [i[0]] + replaceDots(i[1..])
+      assert replaceDots(i) == [i[0]] + replaceDots(i[1..]);
+      if i[0] == '_' {
+        assert |i| >= 2 &&
+               i[1] in "_qkh" && isDafnyEncodedIdTail(i[2..]);
+        replaceDotsReversible(i[2..]);
+        assert reverseReplaceDots(replaceDots(i)) == i;
+      } else {
+        replaceDotsReversible(i[1..]);
+        assert reverseReplaceDots(replaceDots(i)) == i;
+      }
+    }
+  }
+
+  // The reason why escapeIndent is correct is that
+  // - It receives either
+  //     - a Dafny-encoded Id that has no special character (only letters + "__" which came from "_")
+  //       and is not a reserved Rust keyword
+  //       ==> Use idiomatic rust
+  //     - a is_tuple_numeric like _<1 or 2 digits>
+  //       ==> Use idiomatic rust
+  //     - a is_tuple_builder like ___hMake<1 or 2 digits>
+  //     - a dafny-generated identifier that starts with a "_" like "_module"
+  //     - A rust keyword
+  //       ==> use "r#" + rust keyword
+  //     - Anything else
+  //       ==> use "r#_" + replaceDot
+  // - The reverse function is such that
+  //   - If the input is a Dafny-encoded Id or a tuple numeric,
+  //     it can find the original identifier based in the output
+
+  lemma better_tuple_builder_name_not_rsharp(i: string)
+    ensures is_tuple_builder(i) ==>
+              var s := better_tuple_builder_name(i);
+              s[0..2] != "r#"
+  {}
+
+  lemma idiomatic_rust_not_rsharp(i: string)
+    ensures
+      !has_special(i)
+      ==>
+        |idiomatic_rust(i)| < 1 || idiomatic_rust(i)[0] != '#'
+  {
+    if !has_special(i) {
+      var s := idiomatic_rust(i);
+      if |i| == 0 {
+      } else if i[0] == '_' {
+        var s := "_" + idiomatic_rust(i[2..]);
+        assert s[0] == '_';
+      } else {
+        var s := [i[0]] + idiomatic_rust(i[1..]);
+        assert s[0] == i[0];
+      }
+    }
+  }
+
+  lemma {:fuel has_special, 2, 3} escapeIndentExamples()
```

Comment:
Discussion: should we put "test lemmas" outside of the source files, like tests?

Reply:
Based on your discussion topic, I added the following at the top of this file this comment, because I should make it explicit that:
```
/*This module does not contain any compiled code because it
  only proves properties about DCOMP. In a sense, it's a test file. */
```
So you are already viewing here a file that only contain test lemmas. It's just that the others are more like abstract tests, while this one is more concrete.

I could also have made it a method and expect statements to not overload the verifier. Using the verifier to perform computation is not something to advise in general. Here I kept my examples to use the least amount of fuel (the fuel indicated is the minimum + 1) and this can help to design quickly any new bijection feature, and also it makes it possible to not compile this module until we have a way to filter out test modules for compilation for production code.

I also renamed this lemma to `escapeIdentExamples` (fixed the typo)

Reply:
Ah that is illuminating. I had assumed these proofs would be necessary dependencies for future higher-level proofs of correctness, and therefore more like source code. If that becomes true later we can always relocate/reclassify them.